### ISSUE TITLE
Suggestion: Rearranging Image Collection by Image Name

### ISSUE THREAD

**Issue by hepari** [bjorn/tiled#1856]
As a developer of an ongoing project, I have to add new images to an existing collection. 
The problem is that since the editor shows all images arranged by their id instead of the order of image names, my collection becomes messy very quickly once I start adding new objects.

**Example (current)**
id1: stone1
id2: window

**Now, if I add an image named 'stone2':**
id1: stone1
id2: window
id3: stone2

And the system gets lost right away.
That's why I would love to arrange the objects by name. Is it possible?

**Comment by bjorn** [comment]
> That's why I would love to arrange the objects by name. Is it possible?

In this case, the order in the file is irrelevant and you're just looking for an option to order by name when displaying the tileset, right?

Unfortunately right now this is indeed not an option, but we anyway need a better tileset view that is more suitable for image collections. There should be a thumbnail view, where the tiles are displayed in a uniform size and there should also be an option to display the name. This new view should then definitely also be able to display the tiles sorted by name.

> This is a really pressing matter to me, and would be grateful for an answer

It is hard to say when I would get around to this. There are multiple other features on my priority list that I'm currently working on, so it will take at least some weeks. Of course, maybe somebody else would be interested in implementing this.

**Comment by keshav2010** [comment]
I will try to devote some time on this issue  (although my first internal examination is going to start from 19nth of February) i will be digging up the source code and see how much of the requested feature i can implement.

**Comment by hepari** [comment]
@keshav2010 : Thank you for showing interest in this issue. I actually meant the roster of objects under the 'tilesets' window! Right now, they are always sorted by id numbers which are assigned in the order of being added to the tileset file.

[ID-01] [ID-02] [ID-03]
[ID-04] [ID-05] ...

But I want them to be sorted automatically by their file name, like in the following example!

![capture-20180207001](https://user-images.githubusercontent.com/10383640/35944171-d276fbbe-0c5b-11e8-888c-1e23c092ee26.png)

I'd be really, really grateful if you could implement this one.

**Comment by keshav2010** [comment]
Oh i see, so you basically want to re-arrange those individual tiles in a given Tileset.
From the build I've on my system, currently their isn't by default any Name property _**for each tile**_ in a Tileset and although you can sure add a Custom property for name but that wouldn't be really helpful since you might not be willing to assign a name to all tiles in a given tileset. 

I will first make sure every tile in a tileset must have a name-property, which by default is same as their ID and so every Tile should then be sorted on the basis of this Name property only.

**Comment by hepari** [comment]
@keshav2010 : OK! Could you set the name-property as same as each tile's file name? I forgot to mention that this is a problem with the '**Image Collection**', and not the traditional '**tileset**'. I really appreciate your help!!

**Comment by keshav2010** [comment]
Yup name-property could be set to tile's file name, Although i never really worked with Collection of Images before from Tiled User perspective, but still from what I've seen so far by experimenting with it on my system, this is still similar to what i thought about earlier. I will dive into this asap. 😃

**Comment by hepari** [comment]
@keshav2010 : Hello Keshav, how is it going? I hope it didn't turn out to be much more complicated than thought.

**Comment by keshav2010** [comment]
@hepari Hello, Yea i did a little bit of sneak peak into the code today and got started with the implementation but still alot of work is to be done, right now im working on adding name property to each tile and inspecting source code to implement this basic fundamental property safely like other existing properties. It will take time also because I've exams coming up from 19nth Feb and they are expected to last 4 to 5 days, so that is another reason why I'm not able to devote much time to this issue. 

Rest assured, I've no other issue in my to-do list except this one. 😄  

Also here's the short review of what is being done right now, im still working on making sure the name property works properly, i added code to make sure tile name by default is same as tile's file name in case if it is image collection, and so **any change in file name should reflect back in name property,** so these little actions need to be handled before i actually move on to the main problem of sorting them. 
To sum it up, right now im exploring source code and understanding all these little bit of possibilities that can affect name property later in the program.

**Comment by hepari** [comment]
@keshav2010 : oh, I didn't mean to rush you or anything :) Thank you for giving me some insight into your work! 
I assume that without your help, I would have to wait months (or years, in the worst case) until this issue gets addressed. Thank you very much! 😃 

![capture-20180117007](https://user-images.githubusercontent.com/10383640/36066921-62a2a824-0eb3-11e8-974d-6829f73126a0.png)

**Comment by bjorn** [comment]
The most tricky bit here will be to implement a new tileset view, probably based on `QListView`, and the ability to switch between that view and the current `QTableView` based one. Apart from sorting, such a view would offer additional advantages:

* Displaying each tile with the same size (or at least it could be an option). That would solve the layout issues where much space is wasted because of a single larger tile.

* Dynamically wrapping the tiles instead of having to set a "column count" property somewhere.

> I assume that without your help, I would have to wait months (or years, in the worst case) until this issue gets addressed. Thank you very much! 😃

This is an issue I'd want to get around to sooner rather than later, but of course you may be right. It's mostly since I'm just one person and there are so many things to do.

**Comment by hepari** [comment]
@bjorn : of course! I hope that didn't read like a bash directed at you. I'm grateful to you for making Tiled, and I'm fully aware of how busy you are. There is only so much one man can do.

**Comment by keshav2010** [comment]
@bjorn  Thank you for valuable feedback, I've added the Name property in the property-dock widget and for _**non Image-Collection**_ tilesets name is equal to ID however for **_Image-Collections based tilesets_**, it is by default equal to the file name. As of now, **I've not made name property mutable** **so user won't be able to change it directly** **but only by changing file name**, although this is done only to keep things simpler at the initial stage as I'm not sure if making it mutable would mean going more deeper into the code, which at the moment im trying to avoid (due to exams).

**Comment by hepari** [comment]
@keshav2010 : I don't think we'd need to make the property name mutable, as long as we can replace the file! :)

**Comment by hepari** [comment]
Hello @keshav2010 , have you made any progress on this feature? :)

**Comment by keshav2010** [comment]
@hepari  Hey, I've been working on implementing the new tileset view lately based on QListView as suggested by @bjorn however due to quite a hectic schedule I'm not able to devote much time to it, though I'm making progress however its still slow, I will try to gear up the progress as soon as possible. As I'm working on this, I'm learning new things side by side about Qt and Tiled code organization and even though I might not able to meet coding standards being followed, I will make sure to update my own personal fork Tiled with the changes you requested so you can use the new Tileset View without waiting for it to come to official Tile release.😃 
**_A note to Bjorn and anyone else interested in this Issue_**
I hope I'm not blocking anyone's motive to work on this Issue. if anyone is willing to work on this, they should do it ❤️  , I just kind of feel like I might take a bit longer to implement this and this might be causing others to think of this issue as being completely reserved to me. ❤️

**Comment by hepari** [comment]
@keshav2010 Thank you very much for all your effort! Please let me know when you have updated a new fork, I'm eager to try it out :D I'm so gonna credit you in the 'Special Thanks to' section of our game...

**Comment by keshav2010** [comment]
@hepari  I've recently pushed a very small alternate solution to the problem, Since as I said before, implementing whole new tileset view would take some time since I've to follow code guidelines as well as understand various other aspects, I've done some changes in tilesetModel instead and what it basically does, it will internally store data sorted based on its Name property rather than ID. Note that it will only be affected for image collections and not general tileset since as we discussed earlier. There sure are some problem that you might face since they are currently sorted based on character-by-character ASCII code comparison, so that means in a collection,  for example, file namely "9" is considered greater than file namely "10", although I've made sure such behavior should not be shown for non-collection tilesets and they work as expected. I won't be pushing this on Tiled's official repository since this isn't actually a strong fix as it still does not solve issue of extra spacing as mentioned by Bjorn, so if in case it is very urgent for you, you can see my fork of Tiled and instead build it to try it out: https://github.com/keshav2010/tiled 

Meanwhile, I will be work on the dedicated tileset view and rather not alter the existing one. 😄  ❤️

**Comment by hepari** [comment]
@keshav2010 Thank you so much.  It sounds like your build will solve my problem - though I don't know how to compile the build. (I've downloaded qt and opened the unzipped tiled.qbs, but building process just spits some error messages and I don't know what I'm doing wrong XD)

![capture-20180318004](https://user-images.githubusercontent.com/10383640/37565191-c9b84ebc-2aa4-11e8-87c7-5567dd37bbca.png)

Could you help me with this, or give me a link to a compiled build? I'm really sorry for the hassle. :(

**Comment by keshav2010** [comment]
Yea, you can find me on facebook and we can use team viewer to resolve issues and build Tiled on your system : https://www.facebook.com/keshav15157

**Comment by hepari** [comment]
@keshav2010 : Thank you for your friendly offer! Though I don't use Team Viewer ever since, like, ever. I'll just try my luck some other way. 'Readme' is kept so simple and presumes that I know everything about the witchery of compiling...

**Comment by keshav2010** [comment]
No problem, do let me know if the fork solved the issue for you since it is working fine so far. If you face trouble, I will try to build tiled from my fork and provide you with the download link.

**Comment by hepari** [comment]
@keshav2010: Thank you very much, our programmer compiled it for me, and at first it gave us the error message that the app can't find qwindows plugin, but we got that one solved.
The objects sorting is working as I wanted it to, and it hasn't made any boo-boo so far!
I must say, this is already a big improvement over the vanilla version. Thank you! I'll happily wait for the rest of the features, and hope maybe @bjorn could incorporate this change to the master branch so everyone can benefit from it :)

**Comment by hepari** [comment]
@bjorn  @keshav2010 
Bjorn, would you consider adding the codes written by keshav to the main fork? I have been using his version, and the object sorting by name has helped me immensely. I haven't faced any problem using it so far, and can't imagine doing without that feature. 
Having tested it for good 8 months now, I'm 100% sure that other people could only benefit from it.

**Comment by bjorn** [comment]
@hepari It would be best for @keshav2010 to open a pull request with those specific changes then. That makes it is easier to discuss his solution and the possibility of merging it to master.

**Comment by keshav2010** [comment]
@hepari  @bjorn  hey guys 😃 , been away from this issue for a long time and @hepari  glad to hear that the sorting solution survived for 8 months 😄 . I will open the pull request, although the solution is, I believe, not worth it, still, I will open PR and let @bjorn suggest any changes. I will open PR within 1-2 days. 😺

**Comment by hepari** [comment]
@keshav2010 Thank you very much! It's such a huge quality-of-life improvement. Too bad I'm not a coder... I've got a bunch of ideas to make the tool better.

**Comment by zachmoss1** [comment]
I am also interested in this issue, and would add that, ideally for me at least, the tiles from a tile collection in the tileset viewer could be organized arbitrarily by the user. There may be options to organize by ID, name, filename or in an arbitrary manner that is linked to none of those so that tiles can be visually organized (via drag and drop) by the user from within tiled.

**Comment by bjorn** [comment]
> Too bad I'm not a coder... I've got a bunch of ideas to make the tool better.

You're welcome to open more issues with suggestions!

**Comment by bjorn** [comment]
I've now added a dynamic wrapping mode, but I did it by adjusting the column count based on the width of the view (as well as displaying all tiles at the same size, otherwise we can't reliably calculate the amount of visible columns). So, since it's not a list view we can't easily implement various sorting methods, though it can be added like in PR #2046.

**Comment by hepari** [comment]
> I've now added a dynamic wrapping mode, but I did it by adjusting the column count based on the width of the view (as well as displaying all tiles at the same size, otherwise we can't reliably calculate the amount of visible columns). So, since it's not a list view we can't easily implement various sorting methods, though it can be added like in PR #2046.

Thank you for your hard work! So rearranging by names isn't implemented yet, correct? I'd be ecstatic if you could add it too :)

**Comment by Arcanorum** [comment]
Was this ever done? I also have this desire to rearrange or otherwise have some kind of order to the items in a collection of images tileset.

**Comment by hepari** [comment]
> Was this ever done? I also have this desire to rearrange or otherwise have some kind of order to the items in a collection of images tileset.

Sadly, the answer is no... Despite all the great quality of life improvements, this crucial feature has still not yet been implemented. Which means I'm still hopelessly stuck with the customized version. 
![image](https://user-images.githubusercontent.com/10383640/94561895-efe9e380-0264-11eb-807c-e4dc0896acd8.png)

**Comment by bjorn** [comment]
@hepari @Arcanorum I hear you and will have a look at including this in the upcoming new feature release.

**Comment by bjorn** [comment]
I'm sorry this didn't make it into the Tiled 1.5 release! The good news is, that there is a related pull request #2983 by @jmi2k that we'll try to finish this month, so it shouldn't take much longer now.

**Comment by hepari** [comment]
That's great to hear. Tiled has become much more user-friendly with the
recent releases. :)
I would be immensely grateful if it would at least make it into the
snapshot build so that I can start making use of it right away.

Thank you for your hard work!

**Comment by jmi2k** [comment]
Currently it still has some bugs. If you are willing to compile it, I can try solving conflicts so you can use the latest version, but I would recommend waiting until it's properly tested and reviewed.

**Comment by hepari** [comment]
I see. I'm in the middle of map making these days, so it would actually
help if I could put that feature to use right now, as long as the bugs
aren't anything too serious. :)
